src/frontend/context/SecurityContext.tsx

```diff
@@ -0,0 +1,155 @@
+import * as React from "react";
+
+import { OBSCURE_KEY, OBSCURE_PASSCODE, PASSWORD_KEY } from "../constants";
+import createPersistedState from "../hooks/usePersistedState";
+
+type AuthState = "unauthenticated" | "authenticated" | "obscured";
+
+type AuthSetters =
+  | { type: "passcode"; value: string | null }
+  //This is to set up the future use of the obscure pass beings set by the user. if the `value` is undefined, the default obscure pass is being used
+  | { type: "obscure"; value?: string | null };
+
+type AuthValuesSet = {
+  passcodeSet: boolean;
+  obscureSet: boolean;
+};
+
+type SecurityContextType = {
+  authValuesSet: AuthValuesSet;
+  setAuthValues: (val: AuthSetters) => void;
+  authenticate: (val: string | null, validateOnly?: boolean) => boolean;
+  authState: AuthState;
+};
+
+const DefaultState: SecurityContextType = {
+  authValuesSet: { passcodeSet: false, obscureSet: false },
+  setAuthValues: () => {},
+  authenticate: () => false,
+  authState: "unauthenticated",
+};
+
+export const SecurityContext = React.createContext<SecurityContextType>(
+  DefaultState
+);
+
+const usePersistedPasscodeState = createPersistedState(PASSWORD_KEY);
+const usePersistedObscureState = createPersistedState(OBSCURE_KEY);
+
+export const SecurityProvider = ({
+  children,
+}: {
+  children: React.ReactNode;
+}) => {
+  const [authState, setAuthState] = React.useState<AuthState>(
+    "unauthenticated"
+  );
+
+  const [passcode, passCodestatus, setPasscode] = usePersistedPasscodeState<
+    string | null
+  >(null);
+
+  const [obscureCode, , setobscureCode] = usePersistedObscureState<
+    string | null
+  >(null);
+
+  // if passcode is unset, we want to makes sure that obscure code is unset, and that the user is 'authenticated' aka does not need a passcode to enter the app.
+  React.useEffect(() => {
+    if (passcode === null) {
+      setobscureCode(null);
+      setAuthState("authenticated");
+    }
+  }, [passcode]);
+
+  const setPasscodeWithValidation = React.useCallback(
+    (passcodeValue: string | null) => {
+      if (passcodeValue === OBSCURE_PASSCODE) {
```

Comment:
```suggestion
      if (passcodeValue === obscureCode) {
```
Future-proof in case we forget later

Reply:
Obscure code will ALWAYS be set after the passcode. This is only a check that for when we have the default one. I will add a to do comment here

Reply:
What about the following flow?

1. User sets a passcode
2. User sets an obscureCode
3. User changes their passcode

Reply:
That is what this [useEffect here](https://github.com/digidem/mapeo-mobile/pull/1013/files/6599a58d695a8f81a137cf915016026f37dbdce9#diff-476d432624050bbf2aca57f2cd9260583394f372946e24474c4994df203c8d16) was for, to tie the state of the obscure passcode to the passcode. In otherwords, when the passcode is set to null, the obscure code is set to null.

Reply:
Update here: https://github.com/digidem/mapeo-mobile/blob/77f121f4c65557fa11e41276af0a14d928df0f68/src/frontend/context/SecurityContext.tsx#L79